tornettools/generate_tgen.py

```diff
@@ -434,15 +434,22 @@ def __get_tgen_clients(args, n_users, n_circuits, measurement2):
     # each client will be placed in a country
     country_codes, country_probs = __load_user_data(args)
 
-    # the number of circuits each tgen creates every 10 minutes.
-    # we adjust by the load scale here instead of on n_tgen, because n_tgen
-    # can already be adjusted with args.process_scale.
-    n_circuits_per_tgen = __round_or_ceil(n_circuits / n_tgen * args.load_scale)
-
     # split the tgen clients between onion-service clients and non-onion-service (public) clients
     n_hs_tgen = math.ceil(n_tgen * args.onion_service_traffic_scale)
     n_public_tgen = n_tgen - n_hs_tgen
 
+    # the number of circuits each tgen creates every 10 minutes.
+    #
+    # we adjust by the load scale here instead of on n_tgen, because n_tgen
+    # can already be adjusted with args.process_scale.
+    #
+    # Since we only have a measurement of the number of exit circuits (not hidden service circuits),
+    # we calculate load by distributing those over the public (not hiddens ervice) clients.
+    # 
+    # For now we apply the same circuit load to the hidden service clients,
+    # though it'd perhaps be preferable to measure and model them separately.
+    n_circuits_per_tgen = __round_or_ceil(n_exit_circuits / n_public_tgen * args.load_scale)
```

Comment:
I think there is a problem with doing it this way. Suppose someone configures 
  - `onion_service_traffic_scale=0.99`

Then we'll expect the remaining 1% of clients to create all of the `n_exit_circuits` giving us a very large `n_circuits_per_tgen`, and the very large circuit creation rate will then also get used by the HS clients. We'll end up overcompensating and creating too much load.

How about instead, we:
- Estimate that HS traffic accounted for roughly 5% of usage
- Instead of using `args.onion_service_traffic_scale` to set the HS/public split, we instead use it as a multiplicative modifier on the number of HS clients, kinda like we do with `load_scale` for circuits.
  - Here 1.0 means "the normal number of HS clients" and then you can adjust that if you want more or fewer clients.
  - We may want to rename `onion_service_traffic_scale`? and update the help menu docs.

So line 438-439 becomes something like
```python
# Estimated traffic fraction during measurement. When we update
# our measurements, we should get a more precise number here.
estimated_hs_frac = 0.05
n_hs_tgen = math.ceil(n_tgen * estimated_hs_frac)
n_public_tgen = n_tgen - n_hs_tgen

# Apply user-specified adjustments to the default onion service client count
n_hs_tgen = __round_or_ceil(n_hs_tgen * args.onion_service_traffic_scale)
```

Then we compute `n_circuits_for_tgen` the same way, ie
```python
n_circuits_per_tgen = __round_or_ceil(n_exit_circuits / n_public_tgen * args.load_scale)
```
and that is used to create circuit creation models for both exit and hs clients.

There is a question of if we want the default for the new `args.onion_service_traffic_scale ` to be 1.0 or 0.0; I think we could go ahead with 1.0 unless you think that would cause too many problems.

Reply:
Or maybe we'd want to set the new `args.onion_service_traffic_scale` to 0.0 for now to match the current state, and then adjust the default later after we get some more experience with using it.

Reply:
Done. I ended up reworking this a bit.

* Rearranged things and introduced some intermediate variables to avoid assigning one value to a variable and then mutating it later. i.e. try to have a variable names mean the same thing throughout.
* Used "exit" instead of "public" for non-hidden-service clients etc. This avoids confusion with references to the "public tor network". It does overload "exit" a bit, but I think it's clear in context.
* Apply the minimum client threshold to each type separately. Before I was just applying it to the total, which could leave us with a very small number of hidden service clients.
* Updated logging and parsing to record the breakdowns of hidden-service vs exit users, tgens, etc. (Also changed the parsing to use regexes, which will hopefully be a bit easier to understand and maintain)
* Filtered out onion service metrics in the tgen parsing step, since we don't want them lumped together. I'll add parsing and plotting for each individually in a follow-up.